project: add CONTRIBUTING.md and issue templates (#558)
twilight-rs/twilight: CONTRIBUTING.md

# Issues

Issues have three types: bug, feature request, and support. When reporting a bug, you must include
the operating system used, any relevant information about the tech stack, and the feature flags

> used. 

Author: I don't have any policy in mind for feature requests or support issues. Please suggest changes.

Reviewer: > support issues

We should probably try redirecting people to the Discord for support and leave support issues as either a tracking issue or fallback if the person is opposed to using Discord. For the issue contents though, I think the operating system and feature flags are solid. We could also attempt to ask for a reproduction example to make Twilight dev's lives easier. The reporter must also provide whatever error rustc is giving them, if applicable. 

> feature requests

I'm thinking of a format like this:
```markdown
## What would you like implemented?
<Thing they want that Twilight lacks>

## What use cases does this request address?
<More then a single project's specialized use case>

## Are you willing to help towards contributing this feature?
Note: Its perfectly fine if you say no, but we encourage you to try :) 
<The submitter's answer> 

## What would the feature look like and other info
<Other snippets, examples, etc>
```

This would deflect the low effort "Give me xyz" issues every project dreads and, I think at least, sits on a nice sweet spot of detailed and too much work.

# Pull Requests

Pull requests should be named with the format `crate: short description of change`, and should use

> lower case letters. Always make a pull request from a new branch, that is named similarly, but with

Reviewer: > ... and should use lower case letters

Does this include proper nouns like Discord, Redis, etc?

Author: I think so, yes.

Reviewer: I'm not a fan, but I'll let others throw their opinions into the hat on this one.

> lower case letters. Always make a pull request from a new branch, that is named similarly, but with

Reviewer: Should we really enforce branch names? Seems harsh. I think just `Always make a pull request from a feature branch with a related title` is enough and stops the main issue of people working off `trunk` on their forks.

Author: Up for discussion, I'm fine with either, but we should start denying PRs from trunk

Reviewer: I'd say just go for banning PRs from trunk so its not too strict but stays reasonable. Branch names can be fun to title sometimes :)

Reviewer 2: i don't understand this restriction.

Author: If someone PRs trunk->trunk, it is hard to pull in the proposed changes into the same folder and dev environment as one's own fork, making testing harder. One must clone the entire fork into a new directory to test. If new branches are made, and with a relatively unique name, it's easier to add the contributor's fork as a remote and pull in changes that way.

Reviewer 2: You can change the local name with `git checkout -b <branchname> --track upstream/<branchname>`.

Author: I didn't know this, thanks. Still want to enforce this rule.

only a few words: `crate-short-description`. If adding a feature or enhancement, use the term `add`
or something sufficiently similar. If fixing a bug, use the term `fix`, or something sufficiently

> similar. Avoid force-pushing to a pull request branch, as this erases review comment history.

Reviewer: Its probably worth noting here that the default policy is to squash on merge, otherwise this will appear quite odd that we encourage tons of commits.

Author: That is noted in the section about merging

Reviewer: (Probably by linking to the `# Merging` section since its listed there)

# Labeling

> If you can, you must label your issues and pull requests appropriately. This includes adding a label

Reviewer: This is contradictory saying "can" and then following it with "must" imo. It also doesn't define what "can" means.

```suggestion
If an applicable labels exist, then you must label your issues and pull requests appropriately. This includes adding a label
```

Author: I meant that if you are part of the org. I can reword it to `If you are able`

Reviewer: That sounds good to me.

for each applicable crate, or if the issue/change is project-wide, using `c-all`. `feature`s are new
additions, and they are distinct from `enhancement`s, which are improvements on existing features.
`bugfix`es are self-evident. Any change relating to documentation must use the `docs` label. The
`discord api` label is used for changes that must be verified against the Discord API for
correctness.

# Merging

Pull requests require two approvals before merging. They must be merged with the name format `crate:
short description of change`, and they should use lower case letters. The only possible merge option
is squash and merge. Add the headers `Approved-by`, `Merged-by`, and `Signed-off-by` (the author of
the pull request) to the end of the merge description. These headers must contain data in the format

> `Name <email@example.com>`, unless that data is not present.

Author: Should these be required? We stopped adding them purely because of laziness.

Reviewer: I think its good to keep some kind of audit trail, but it also discourages potential contributors from merging since this practice has been dying out. Twilight is the only project on GitHub I've seen care about these.

Author: I suppose the data does exist in the PR itself, which is linked in the commit when browsing on Github. I think that was the reason we didn't start doing them again last time this came up

Reviewer: Worst case: We move off GitHub for some reason and lose the past links.

Reviewer 3: Could be noted that the only part needed to be addded by the person making the pr is to add `-s`/`--signoff` to the `git commit` command.

# Issues

Issues have three types: bug, feature request, and support. When reporting a bug, you must include
the operating system used, any relevant information about the tech stack, and the feature flags
used. Feature request issues should answer these questions: 

```
## What would you like implemented? What do you want that Twilight lacks?

## What use case does this request address? This should be more than a single project's use case.

## Are you willing to help towards contributing this feature?

## Is there any other information that we should know?

> ```

Reviewer 3: Should this just not be a markdown list?

Reviewer 4: Most probably
```suggestion
- What would you like implemented? What do you want that Twilight lacks?
- What use case does this request address? This should be more than a single project's use case.
- Are you willing to help towards contributing this feature?
- Is there any other information that we should know?
```

Before making a support issue or bug report, be sure to consider joining the
[Discord](https://discord.gg/7jj8n7D) and bringing up your topic in the #support channel.

# Pull Requests

Pull requests must be named with the format `crate: short description of change`, and should use
lower case letters. If the change spans more than one crate, separate the crate names with a comma
and a space: `crate1, crate2: short description of change`. Pull requests must be made from a new

> branch. If a PR is made from the `trunk` branch, it will be denied. If adding a feature or

Reviewer 3: I don't think we should flat out deny them, rather ask to avoid it maybe

enhancement, use the term `add` or something sufficiently similar. If fixing a bug, use the term
`fix`, or something sufficiently similar. Avoid force-pushing to a pull request branch, as this

> erases review comment history.

Reviewer 4: Possibly add a section on when members should merge pull requests. ie. right after the 2 approvals

# Labeling

If you are able, you must label your issues and pull requests appropriately. This includes adding a
label for each applicable crate, or if the issue/change is project-wide, using `c-all`. `feature`s
are new additions, and they are distinct from `enhancement`s, which are improvements on existing
features.  `bugfix`es are self-evident. Any change relating to documentation must use the `docs`
label. The `discord api` label is used for changes that must be verified against the Discord API for
correctness.

# Merging

> Pull requests require two approvals before merging. They must be merged with the same name as the

Reviewer 3: Usually the name is `{pr_name} (#{pr_number})`

pull request, and as such must follow the naming rules for pull requests. The only possible merge

> option is squash and merge.

Reviewer 4: We should add a section on the structure of the repository. ie. the difference between crates in tree and those in a separate repository

```suggestion
option is squash and merge.

# Adding new crates
When a crate is in development or planned to be added, it will not be created in a pull request since that would make it harder to contribute to until it is merged. Instead, new/in development crates will be placed in their own separate repository and will be merged in-tree once the crate is feature complete, tested, and to the standards of the rest of the ecosystem.

# Testing
When changes are added, tests should be added to ensure no regressions will erupt in the new code. It is crucial when implementing a bug-fix that tests are created to ensure that the crate cannot regress and to prevent a new bug from forming from the fix.

# Documentation
When adding new features, or changing an existing user-facing api, documentation should be created or updated respectively and should contain code examples when possible
```

Author: I'm not sure about the new crates policy. If I recall correctly, adding new crates has been a case-by-case thing. Lavalink was added with a PR, I'm not sure what happened with the new util crate stuff, and the Oauth crate was a PR but then was moved to a repository due to changes in priorities.

Reviewer 4: I think we should standardize some workflow of adding a crate.

Reviewer 4: Yah, that make sense. I'll resolve this.

Reviewer 4: I have updated the suggestions to remove the info on new crates since it makes sense that there will be standardized flow

# Issues

Issues have three types: bug, feature request, and support. When reporting a bug, you must include
the operating system used, any relevant information about the tech stack, and the feature flags
used. Feature request issues should answer these questions: 

```
- What would you like implemented? What do you want that Twilight lacks?
- What use case does this request address? This should be more than a single project's use case.
- Are you willing to help towards contributing this feature?

> - Is there any other information that we should know?

Reviewer 5: I feel like this is duplication and would just say to use the according templates

/// specified below.
///
/// When documenting struct fields, don't prefix with "The", otherwise most documentation lines
/// would start with "the".
///
/// [named anchors]: https://api.twilight.rs
struct Structy {
    /// Something.
    pub some: String,
    /// Something else.
    pub stuff: String,
}
```

Methods are to be documented as follows:
```rust
impl Structy {
    /// Short decription of the method.
    ///
    /// More important information or clarification.
    pub fn method(&self) -> Option<Something> {

    }
}
```

# Labeling

If you are able, you must label your issues and pull requests appropriately. This includes adding a
label for each applicable crate, or if the issue/change is project-wide, using `c-all`. `feature`s
are new additions, and they are distinct from `enhancement`s, which are improvements on existing
features. `bugfix`es are self-evident. Any change relating to documentation must use the `docs`
label. The `discord api` label is used for changes that must be verified against the Discord API for
correctness.

# Merging

Pull requests require two approvals before merging. The commit must be named with the format

> `{pr name} (#{pr number})`. The only possible merge option is squash and merge.

Reviewer 5: Should probably add some kind of guide for commit messages here

used, as specified in the issue template. Feature requests also have an issue template, containing
questions that should be answered in the issue. We aim for 100% coverage of the Discord API; we will
wait until a new feature is supported in the [Discord API documentation] before adding its support
to Twilight. Before making an issue, be sure to consider joining the [Twilight Discord] and bringing
up your topic in the `#support` channel.

# Pull Requests

Pull requests must be named with the format `{crate}: {short description of change}`, and should use
lower case letters. If the change spans more than one crate, separate the crate names with a comma
and a space: `{crate1}, {crate2}: {short description of change}`. Pull requests must be made from a
new branch. Please avoid making pull requests from the `trunk` branch. If adding a feature or
enhancement, use the term `add` or something sufficiently similar. If fixing a bug, use the term
`fix`, or something sufficiently similar. Avoid force-pushing to a pull request branch, as this
erases review comment history.

Contributors should add tests and documentation that reflects their changes.

## Tests

> Contributors should write tests to ensure soundness in present and future code. Alongside regular

Reviewer: As noted in Discord, this should be `correctness` and not `soundness` since we aren't writing any unsafe :eyes:

struct Structy {
    /// Something.
    pub some: String,
    /// Something else.
    pub stuff: String,
}
```

Methods are to be documented as follows:
```rust
impl Structy {
    /// Short decription of the method.
    ///
    /// More important information or clarification.
    pub fn method(&self) -> Option<Something> {

    }
}
```

# Labeling

If you are able, you must label your issues and pull requests appropriately. This includes adding a
label for each applicable crate, or if the issue/change is project-wide, using `c-all`. `feature`s
are new additions, and they are distinct from `enhancement`s, which are improvements on existing
features. `bugfix`es are self-evident. Any change relating to documentation must use the `docs`
label. The `discord api` label is used for changes that must be verified against the Discord API for
correctness.

# Merging

Pull requests require two approvals before merging. The only possible merge option is squash and
merge. The commit must be named with the format `{pr name} (#{pr number})`. When merging, add
headers to the commit message that show who approved, merge, and authored the commit. The header

> `Signed-off-by` is used to specify the commit author. Refer to [this example commit] for proper

Reviewer 4: We should also mention at least `Merged-by` for use by the merger and possibly also `Approved-by` for approvals.

Author: I believe it is specified in that selection, but I can make it more clear.
